### PR TITLE
chore(kuma-cp): improve logging in K8S controllers

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
As a follow-up of https://github.com/kumahq/kuma/pull/4964#discussion_r965701767 and conversation with @michaelbeaumont offline.
I went through all controllers and added missing logs for significant events that change the system state.

### Checklist prior to review

- [ ] Link to docs PR or issue --
- [ ] Link to UI issue or PR --
- [ ] Is the [issue worked on linked][1]? --
- [ ] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [ ] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Unit Tests --
- [ ] E2E Tests --
- [ ] Manual Universal Tests --
- [ ] Manual Kubernetes Tests --
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [ ] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch)?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_controller.go [hunk 1]
```diff
@@ -52,7 +52,7 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request
 		if kube_apierrs.IsNotFound(err) {
 			// We don't know the mesh, but we don't need it to delete our
 			// object.
-			err := common.ReconcileLabelledObject(ctx, r.TypeRegistry, r.Client, req.NamespacedName, core_model.NoMesh, &mesh_proto.MeshGateway{}, nil)
+			err := common.ReconcileLabelledObject(ctx, r.Log, r.TypeRegistry, r.Client, req.NamespacedName, core_model.NoMesh, &mesh_proto.MeshGateway{}, nil)
```
Comment (lahabana) [inline pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_controller.go:55]: Does it make sense to pass in a log already init with `namespace` and `name` (and maybe type) as the `reconcile` log already almost logs this.

File: pkg/plugins/runtime/k8s/plugin.go [hunk 2]
```diff
@@ -145,6 +145,7 @@ func addMeshReconciler(mgr kube_ctrl.Manager, rt core_runtime.Runtime, converter
 	}
 	defaultsReconciller := &k8s_controllers.MeshDefaultsReconciler{
 		ResourceManager: rt.ResourceManager(),
+		Log:             core.Log.WithValues("controllers").WithName("mesh-defaults"),
```
Comment (lahabana) [inline pkg/plugins/runtime/k8s/plugin.go:148]: Should there be a value with the `controllers` key?

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/common/reconcile.go [hunk 3]
```diff
@@ -81,11 +90,20 @@ func ReconcileLabelledObject(
 	}
 
 	if existing != nil {
+		existingSpec, err := existing.GetSpec()
+		if err != nil {
+			return err
+		}
+		if proto.Equal(existingSpec, ownedSpec) {
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/k8s/controllers/gatewayapi/common/reconcile.go:97]: kind of sneaking an unrelated change in here :wink: 

This happens all over the place. And it seems error prone, what if the meta has changed? Can we do this instead with `client.Patch`?

Comment (jakubdyszkiewicz) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/common/reconcile.go:97]: Do you mean that meta can change in other places? From what I'm reading, `ReconcileLabelledObject` retrieves an existing object and updates the spec.

Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/common/reconcile.go:97]: I just meant it's fragile. If logic is ever introduced earlier on that does change meta, this would be broken. 

Of course this is only one spot but before we add `proto.Equal` guards inline before k8s `Update` calls everywhere we can probably as a rule use a helper that works with `Patch` and skip the API call if the patch is empty. I'll look into it

File: pkg/plugins/runtime/k8s/plugin.go (hunk 2 above)
Comment (jakubdyszkiewicz) [reply pkg/plugins/runtime/k8s/plugin.go:148]: Could you elaborate?

Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/plugin.go:148]: I think it's `WithValues(key, value interface{})`

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/common/reconcile.go (hunk 3 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/common/reconcile.go:97]: Discussed offline, we'll merge this for now and come up with a helper if this pattern is necessary elsewhere.

File: pkg/plugins/runtime/k8s/plugin.go (hunk 2 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/plugin.go:148]:
```suggestion
		Log:             core.Log.WithName("controllers").WithName("mesh-defaults"),
```

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_controller.go (hunk 1 above)
Comment (jakubdyszkiewicz) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_controller.go:55]: I'm not sure. We always pass this data so I guess it's easier to use `ReconcileLabelledObject` if you can just pass the log without extra line?

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/common/reconcile.go (hunk 3 above)
Comment (jakubdyszkiewicz) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/common/reconcile.go:97]: Talked offline, we decided to leave it for now as this is a helper already. If we see more occurrences of such logic, we can try to extract this to a helper.

File: pkg/plugins/runtime/k8s/plugin.go (hunk 2 above)
Comment (jakubdyszkiewicz) [reply pkg/plugins/runtime/k8s/plugin.go]: it's actually WithName.WithName. Fixed

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_controller.go (hunk 1 above)
Comment (lahabana) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_controller.go:55]: Ok not strong opinion here and you are the owner here :)

**Review by michaelbeaumont — APPROVED** [review]
